### ISSUE TITLE
Truncated backdrop with max pooling over time

### ISSUE THREAD

**Issue by allentran** [tensorflow/tensorflow#143]
It seems like the way truncated backprop is done, at least in the tutorial on RNNs, is to 
- set up the data in batches where the time dimension is capped at the max `num_steps` 
- run forward/backward passes making sure to use the correct initial state (e.g final state of the previous batch if was part of the same observation but prior in time or the true initial state if the current batch starts at true `t=0`)

This works fine if each hidden state contributes directly to the loss but this approach fails if you want to do something like max pooling across time.  You can carry around the current max (like the correct initial state for truncated backprop) but it seems like the gradient won't flow from the loss, to the max'd elements.  

To summarize, truncated backprop won't work in situations when a given hidden state at time `t`'s contribution to the loss is unknown without knowing the hidden states for all `t`.  

Would love to be wrong.

**Comment by ludimagister** [comment]
No, you are correct. Truncated BP works well for when whatever you are optimizing happens (roughly) in the currently unrolled window. With partial unrolling (compared to complete unrolling) you can do more updates over time, which usually improves convergence speed. Obviously you don't want to unroll for too few steps, as that will increase the error during back-propagation too much. For your problem I would recommend to fully unroll (just increase the num_steps to your maximum sequence length).

**Comment by allentran** [comment]
Alright, makes sense.  Thanks for confirming.
